internal/timex/timextest/fake_clock.go

```diff
@@ -0,0 +1,168 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package timextest provides fake controlled time for testing code using the [timex] package. All methods here are safe
+// to call concurrently.
+package timextest
+
+import (
+	"sync"
+	"time"
+
+	"go.temporal.io/server/internal/timex"
+)
+
+type (
+	// FakeClock is a timex.Clock. Unlike other fake clock implementations, the methods of this clock are synchronous, so
+	// when you call Advance or Update, all triggered timers from AfterFunc will fire before the method returns, in the same
+	// goroutine.
+	FakeClock struct {
+		mu     sync.RWMutex
+		now    time.Time
+		timers []*fakeTimer
+	}
+
+	// fakeTimer is a fake implementation of [timex.Timer].
+	fakeTimer struct {
+		// link to the parent fake time source
+		clock *FakeClock
+		// deadline for when the timer should fire
+		deadline time.Time
+		// callback to call when the timer fires
+		callback func()
+		// called indicates if the timer has already count
+		called bool
+		// index of the timer in the parent fake time source
+		index int
+	}
+)
+
+// NewFakeClock returns a FakeClock with the current time set to the zero value of [time.Time].
+func NewFakeClock() *FakeClock {
+	return &FakeClock{
+		now: time.Time{},
+	}
+}
+
+// Now return the current time.
+func (c *FakeClock) Now() time.Time {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	return c.now
+}
+
+// AfterFunc return a timer that will fire after the specified duration. It is important to note that the clock is
+// locked while the callback is called. This means that you must be cautious about calling any methods on the clock from
+// within the callback. Doing so will probably result in a deadlock. To avoid this, you may want to wrap all such calls
+// in a goroutine.
+func (c *FakeClock) AfterFunc(d time.Duration, f func()) timex.Timer {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	t := &fakeTimer{clock: c, deadline: c.now.Add(d), callback: f}
+	t.index = len(c.timers)
+	c.timers = append(c.timers, t)
+
+	return t
+}
+
+// Update the fake current time.
+func (c *FakeClock) Update(now time.Time) *FakeClock {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	return c.update(now)
```

Comment:
Meh, just accept the time in the `NewFakeClock` if that's the case and document that a zero'd clock is ready to use on its own, you don't have to call this constructor.

Reply:
Ok, I tried this because it makes sense to me, but I realized that the zero value of `time.Time` is pretty gnarly since you can't go backwards from it, and I think that will cause some really unexpected behavior. For that reason, I'm gonna keep the constructor so that the default value is the unix epoch `time.Unix(0, 0)`.

Reply:
Why not just make the constructor accept a time from a user and let the user choose if they want the epoch?

Reply:
I think most people won't care about the initial time, and the unix epoch is a very reasonable default. They can still choose an explicit initial time by running `.Update`

Reply:
It just saves them having to think about what the initial time would be and then writing `time.Unix(0, 0)` each time